src/backend/tty.rs

```diff
@@ -504,28 +506,54 @@ impl Tty {
         let (drm, drm_notifier) = DrmDevice::new(device_fd.clone(), true)?;
         let gbm = GbmDevice::new(device_fd)?;
 
-        let display = unsafe { EGLDisplay::new(gbm.clone())? };
-        let egl_device = EGLDevice::device_for_display(&display)?;
+        let mut try_initialize_gpu = || {
+            let display = unsafe { EGLDisplay::new(gbm.clone())? };
+            let egl_device = EGLDevice::device_for_display(&display)?;
 
-        let render_node = egl_device
-            .try_get_render_node()?
-            .context("no render node")?;
-        self.gpu_manager
-            .as_mut()
-            .add_node(render_node, gbm.clone())
-            .context("error adding render node to GPU manager")?;
+            let render_node = egl_device
+                .try_get_render_node()?
+                .context("no render node")?;
+            self.gpu_manager
+                .as_mut()
+                .add_node(render_node, gbm.clone())
+                .context("error adding render node to GPU manager")?;
 
-        if node == self.primary_node || render_node == self.primary_render_node {
+            anyhow::Ok(render_node)
+        };
+
+        let render_node = try_initialize_gpu()
+            .inspect_err(|err| {
+                warn!("failed to initialze renderer, falling back to primary gpu: {err:?}");
+            })
+            .ok();
+
+        let allocator = render_node
+            .is_some()
+            .then_some(&gbm)
+            .or(self
+                .devices
+                .get(&self.primary_node)
+                .map(|device| &device.gbm))
+            .map(|gbm| {
+                let gbm_flags = GbmBufferFlags::RENDERING | GbmBufferFlags::SCANOUT;
+                GbmAllocator::new(gbm.clone(), gbm_flags)
+            })
+            .ok_or(anyhow!("no allocator available for device"))?;
+
+        let is_primary_render_node = render_node
+            .map(|render_node| render_node == self.primary_render_node)
+            .unwrap_or(false);
+        if node == self.primary_node || is_primary_render_node {
             if node == self.primary_node {
                 debug!("this is the primary node");
             }
-            if render_node == self.primary_render_node {
+            if is_primary_render_node {
                 debug!("this is the primary render node");
             }
 
             let mut renderer = self
                 .gpu_manager
-                .single_renderer(&render_node)
+                .single_renderer(&self.primary_render_node)
```

Comment:
I think it would be useful to extract this change to use `self.primary_render_node` into its own commit just in case future bisecting needs it

Reply:
sure, done